File: hooks/zoto-session-start.py
#!/usr/bin/env python3
"""Session start hook: nudge when workDir contains more items than threshold."""
import json
import sys
from pathlib import Path

DEFAULT_UNIT = "spec"
DEFAULT_WORK_DIR = "specs/current"
DEFAULT_THRESHOLD = 20
DEFAULT_MESSAGE = (
    "You have ${count} unprocessed ${unitOfWork}s. Consider running /zoto-plan to organize."
)


def _emit_empty() -> None:
    print("{}")


def main() -> int:
    sys.stdin.read()

    root = Path.cwd()
    config_path = root / ".spec-system" / "config.json"
    if not config_path.is_file():
        _emit_empty()
        return 0

    try:
        with config_path.open(encoding="utf-8") as f:
            cfg = json.load(f)
    except (OSError, json.JSONDecodeError):
        _emit_empty()
        return 0

    if not isinstance(cfg, dict):
        _emit_empty()
        return 0

    hooks = cfg.get("hooks")
    nudge_cfg = {}
    if isinstance(hooks, dict):
        raw = hooks.get("sessionStartNudge")
        if isinstance(raw, dict):
            nudge_cfg = raw

    if nudge_cfg.get("enabled", True) is False:
        _emit_empty()
        return 0

    unit_of_work = cfg.get("unitOfWork") or DEFAULT_UNIT
    if not isinstance(unit_of_work, str):
        unit_of_work = DEFAULT_UNIT

    work_rel = cfg.get("workDir") or DEFAULT_WORK_DIR
    if not isinstance(work_rel, str):
        work_rel = DEFAULT_WORK_DIR
    work_dir = Path(work_rel)
    if not work_dir.is_absolute():
        work_dir = root / work_dir

    threshold = nudge_cfg.get("threshold", DEFAULT_THRESHOLD)
    if not isinstance(threshold, int):
        try:
            threshold = int(threshold)
        except (TypeError, ValueError):
            threshold = DEFAULT_THRESHOLD

    message_template = nudge_cfg.get("message") or DEFAULT_MESSAGE
    if not isinstance(message_template, str):
        message_template = DEFAULT_MESSAGE

    count = 0
    if work_dir.is_dir():
        count = sum(1 for p in work_dir.iterdir() if p.is_dir())

    if count > threshold:
        msg = message_template.replace("${count}", str(count)).replace(
            "${unitOfWork}", str(unit_of_work)
        )
        print(json.dumps({"additional_context": msg}, ensure_ascii=False))
    else:
        _emit_empty()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/zoto-spec-system/hooks/zoto-session-start.ts
#!/usr/bin/env node
/**
 * Session start hook: nudge when workDir contains more items than threshold.
 *
 * Reads `.zoto/spec-system/config.yml` (the only supported config path).
 * If the file is missing the hook stays silent; the user must run
 * `/z-spec-init` to scaffold one.
 */

import { existsSync, mkdirSync, readFileSync, readdirSync, renameSync, statSync, unlinkSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import YAML from "yaml";

const DEFAULT_UNIT = "spec";
const DEFAULT_WORK_DIR = "specs/current";
const DEFAULT_THRESHOLD = 20;
const DEFAULT_MESSAGE =
  "You have ${count} unprocessed ${unitOfWork}s. Consider running /z-spec-create to organize.";

function emitEmpty(): void {
  process.stdout.write("{}\n");
}

function loadYaml(path: string): unknown {
  try {
    const raw = YAML.parse(readFileSync(path, "utf-8"));
    return raw ?? {};
  } catch {
    return undefined;
  }
}

function countSubdirs(dir: string): number {
  try {
    return readdirSync(dir).filter((entry) => {
      try {
        return statSync(join(dir, entry)).isDirectory();
      } catch {
        return false;
      }
    }).length;
  } catch {
    return 0;
  }
}

function main(): void {
  // Consume stdin (required by hook protocol)
  try {
    readFileSync(0, "utf-8");
  } catch {
    // stdin may be empty or closed
  }

  const root = process.cwd();

  // Auto-migrate legacy .zoto-spec-system/ → .zoto/spec-system/
  const legacyDir = join(root, ".zoto-spec-system");
  const newDir = join(root, ".zoto", "spec-system");
  if (existsSync(legacyDir) && !existsSync(newDir)) {
    try {
      mkdirSync(dirname(newDir), { recursive: true });
      renameSync(legacyDir, newDir);
      const oldCfg = join(newDir, "config.json");
      const newCfg = join(newDir, "config.yml");
      if (existsSync(oldCfg) && !existsSync(newCfg)) {
        try {
          const json = JSON.parse(readFileSync(oldCfg, "utf-8"));
          writeFileSync(newCfg, YAML.stringify(json), "utf-8");
          unlinkSync(oldCfg);
        } catch {
          renameSync(oldCfg, newCfg);
        }
      }
    } catch {
      /* best-effort; loadConfig will handle the fallback */
    }
  }

  const configPath = join(root, ".zoto", "spec-system", "config.yml");

  const cfg = loadYaml(configPath);
  if (!cfg || typeof cfg !== "object" || Array.isArray(cfg)) {
    emitEmpty();
    return;
  }

  const config = cfg as Record<string, unknown>;
  const hooks = config.hooks;
  let nudgeCfg: Record<string, unknown> = {};

  if (hooks && typeof hooks === "object" && !Array.isArray(hooks)) {
    const raw = (hooks as Record<string, unknown>).sessionStartNudge;
    if (raw && typeof raw === "object" && !Array.isArray(raw)) {
      nudgeCfg = raw as Record<string, unknown>;
    }
  }

  if (nudgeCfg.enabled === false) {
    emitEmpty();
    return;
  }

  let unitOfWork = config.unitOfWork;
  if (typeof unitOfWork !== "string" || !unitOfWork) {
    unitOfWork = DEFAULT_UNIT;
  }

  let workRel = config.workDir;
  if (typeof workRel !== "string" || !workRel) {
    workRel = DEFAULT_WORK_DIR;
  }
  const workDir = resolve(root, workRel as string);

  let threshold = nudgeCfg.threshold;
  if (typeof threshold !== "number") {
    threshold =
      typeof threshold === "string" ? parseInt(threshold, 10) : DEFAULT_THRESHOLD;
    if (isNaN(threshold as number)) {
      threshold = DEFAULT_THRESHOLD;
    }
  }

  let messageTemplate = nudgeCfg.message;
  if (typeof messageTemplate !== "string" || !messageTemplate) {
    messageTemplate = DEFAULT_MESSAGE;
  }

  const count = countSubdirs(workDir);

  if (count > (threshold as number)) {
    const msg = (messageTemplate as string)
      .replace("${count}", String(count))
      .replace("${unitOfWork}", String(unitOfWork));
    process.stdout.write(JSON.stringify({ additional_context: msg }) + "\n");
  } else {
    emitEmpty();
  }
}

main();
